### ISSUE TITLE
<globalObjectNames/> Should produce a file containing enum literals for each enum type

### ISSUE THREAD

**Issue by morki** [jOOQ/jOOQ#11375]
### Use case:

With the new JSON support and support for deserialization with Jackson and other libraries, I came across this idea, which will enable us to minimize errors and better use IDE autocompletion.

With Jackson for example, you could use annotatation `@JsonTypeName` on class to tell Jackson the "name" of this class in serialized hierarchies. We use this heavily and the type is defined in database (Postgres) as enum type.

Now, to use this, we need to explicitaly use strings for these names, because in annotations, only compile time constants are supported, and it is very error prone.

```kotlin
@JsonTypeName("TEST")
data class Test(val id: Long)
```

### Possible solution you'd like to see:

It would be cool, if jOOQ's generator can generate those literals as constants in inner class of enum, so we could use then in this situation.

```kotlin
@JsonTypeName(MyJooqGeneratedEnum.Literals.TEST)
data class Test(val id: Long)
```

```java
public enum MyJooqGeneratedEnum implements EnumType {
    // Here the constant literals replaces strings
    TEST(Literals.TEST),

    BLAH(Literals.BLAH),

    private final String literal;

    private MyJooqGeneratedEnum(String literal) {
        this.literal = literal;
    }

    @Override
    public Catalog getCatalog() {
        return getSchema().getCatalog();
    }

    @Override
    public Schema getSchema() {
        return XXX.YYY;
    }

    @Override
    public String getName() {
        return "my_jooq_generated_enum";
    }

    @Override
    public String getLiteral() {
        return literal;
    }

    // This inner class is added
    public class Literals {
        public final static String TEST = "TEST";

        public final static String BLAH = "BLAH";
    }
}
```

### Possible workarounds:

Using manualy typed string.

### Versions:

- jOOQ: 3.14
- Java: -
- Database (include vendor): Postgres
- OS: -
- JDBC Driver (include name if inofficial driver): -

**Comment by morki** [comment]
It also adds "type safety" in the way that when you delete some option from enum in database, the code no longer compiles, and developer must resolve this situation.

**Comment by lukaseder** [comment]
Thanks for your suggestion. This type of enhancement has been suggested a few times in the past, almost always with annotations in mind. If anything, this shows how flawed annotations are, time and again, to achieve what they're doing, but that's an out-of-scope discussion here 😊.

The most relevant duplicate issue (which was rejected) is this one: https://github.com/jOOQ/jOOQ/issues/5210

I understand that people would like to be able to navigate object identifiers of all sorts of generated object types. That is occasionally useful. However, it also adds complexity to an already very complex code generator, with the necessity of:

- Handling the new type of generated artifact in a separate class hierarchy
- Mingling (again) all sorts of naming conflicts
- Allowing for names to be overridden by user generator strategies (both programmatic and configurative)

All this complexity is there for the relative moderate benefit of being able to use generated identifiers in annotations. It is relatively easy to work around the limitation by extending the `JavaGenerator` to generate the additional artifacts.

In the meantime, I will re-open #5210, and put it on hold to collect further user feedback. I recognise that this kind of feature is convenient, but it isn't about a single type of identifier (e.g. enums). It's a significant project that has to be prioritised, and I currently would like to allocate efforts elsewhere, given how simple a custom `JavaGenerator` solution is.

**Comment by superxiao** [comment]
With `globalObjectNames` available in 3.19, would enum literals again become a possibility? It would be useful in Spring `RequestParam.defaultValue` annotations. Happy to create a new feature request if needed. @lukaseder

**Comment by lukaseder** [comment]
Yeah, I guess times have changed. This should be implemented now. No new issue is required.

**Comment by lukaseder** [comment]
> ```java
> // Here the constant literals replaces strings
> ```

For the record, this isn't being done. It unnecessarily complicates the code generation if we always have to check whether to use the generated enum literal (or any other kind of name). The names are simply generated separately for users to use. jOOQ internals won't depend on them.

> ```java
> // This inner class is added
> ```

We're not generating nested classes (let alone inner classes! (though enums can't have inner classes)) as that as well only complicates everything. You can implement your generator strategy to govern package and class names for these additional files, but nesting won't be possible.

**Comment by lukaseder** [comment]
Implemented in jOOQ 3.20.0

**Comment by lukaseder** [comment]
For the record, this is the feature that enables this code generation:
https://www.jooq.org/doc/latest/manual/code-generation/codegen-advanced/codegen-config-generate/codegen-generate-global-object-names/
